clang/runtime/dpct-rt/include/sparse_utils.hpp.inc

```diff
@@ -331,6 +331,524 @@ void optimize_csrsv(sycl::queue &queue, oneapi::mkl::transpose trans,
+  void get_desc(int64_t *row_num, int64_t *col_num, int64_t *nnz,
+                void **row_ptr, void **col_ind, void **value,
+                library_data_t *row_ptr_type, library_data_t *col_ind_type,
+                oneapi::mkl::index_base *base,
+                library_data_t *value_type) const noexcept {
+    *row_num = _row_num;
+    *col_num = _col_num;
+    *nnz = _nnz;
+    *row_ptr = _row_ptr;
+    *col_ind = _col_ind;
+    *value = _value;
+    *row_ptr_type = _row_ptr_type;
+    *col_ind_type = _col_ind_type;
+    *base = _base;
+    *value_type = _value_type;
+  }
+  /// Get the sparse matrix data format of this descriptor
+  /// \param [out] format The matrix data format result
+  void get_format(matrix_format *format) const noexcept { *format = _format; }
+  /// Get the index base of this descriptor
+  /// \param [out] base The index base result
+  void get_base(oneapi::mkl::index_base *base) const noexcept { *base = _base; }
+  /// Get the value pointer of this descriptor
+  /// \param [out] value The value pointer result
+  void get_value(void **value) const noexcept { *value = _value; }
+  /// Set the value pointer of this descriptor
+  /// \param [in] value The input value pointer
+  void set_value(void *value) { _value = value; }
+  /// Get the size of the sparse matrix
+  /// \param [out] row_num Number of rows of the sparse matrix.
+  /// \param [out] col_num Number of colums of the sparse matrix.
+  /// \param [out] nnz Non-zero elements in the sparse matrix.
+  void get_size(int64_t *row_num, int64_t *col_num,
+                int64_t *nnz) const noexcept {
+    *row_num = _row_num;
+    *col_num = _col_num;
+    *nnz = _nnz;
+  }
+  /// Set the sparse matrix attribute
+  /// \param [in] attribute The attribute type
+  /// \param [in] data The attribute value
+  /// \param [in] data_size The data size of the attribute value
+  void set_attribute(matrix_attribute attribute, const void *data,
+                     size_t data_size) {
+    if (attribute == matrix_attribute::diag) {
+      const oneapi::mkl::diag *diag_ptr =
+          reinterpret_cast<const oneapi::mkl::diag *>(data);
+      if (*diag_ptr == oneapi::mkl::diag::unit) {
+        _diag = oneapi::mkl::diag::unit;
+      } else if (*diag_ptr == oneapi::mkl::diag::nonunit) {
+        _diag = oneapi::mkl::diag::nonunit;
+      } else {
+        throw std::runtime_error("unsupported diag value");
+      }
+    } else if (attribute == matrix_attribute::uplo) {
+      const oneapi::mkl::uplo *uplo_ptr =
+          reinterpret_cast<const oneapi::mkl::uplo *>(data);
+      if (*uplo_ptr == oneapi::mkl::uplo::upper) {
+        _uplo = oneapi::mkl::uplo::upper;
+      } else if (*uplo_ptr == oneapi::mkl::uplo::lower) {
+        _uplo = oneapi::mkl::uplo::lower;
+      } else {
+        throw std::runtime_error("unsupported uplo value");
+      }
+    } else {
+      throw std::runtime_error("unsupported attribute");
+    }
+  }
```

Comment:
The design is containing uplo and diag in the descriptor. `cusparseSpMatDescr_t` also contains those attributes. If we use an attribute struct and a descriptor separately, we need insert another variable. This will make a 1:1 mapping to 1:2 mapping which is what we want to avoid.

Reply:
I think this descriptor can also support multiple versions of uplo and diag.
This descriptor just a wrapper of oneapi::mkl::sparse::matrix_handle, if user pass the dense_matrix_descriptor to optimize_trmv with different uplo/diag, oneapi::mkl::sparse::matrix_handle can save the different optimizations, so that the dense_matrix_descriptor also can do that.

Reply:
@spencerpatty is there any further comment?

Reply:
ok, so I completely agree that our mkl::sparse::matrix_handle will keep all the different optimizations you throw at it so they are all usable.  I wonder about your set_optimized_info and is_optimized() function design -- they seem to only be tracking the last optimized entry, unless I am mistaken and am missing some storing mechanism...

Reply:
> they seem to only be tracking the last optimized entry

@spencerpatty , yes, I added this check to avoid calling optimize_xyz() for the same configs. But now I only check the current config and the previous config.
Will mkl::optimize_xyz() do duplicate optimizations if I call it multiple times with same config (same data and same trans/uplo/diag) ?

Reply:
optimize_xyz will check if previous optimizations are already completed and return early (keeping the DAG consistent) if you call it multiple times with same config.  So yes, you are safe.

Reply:
@spencerpatty , then I think I can remove the `is_optimized` check safely.

Reply:
yes this is also a good solution -- just depend on our internal handling of all the opts in the handle